### PR TITLE
keep_kernel option in manifest and is_golden option refactor

### PR DESCRIPTION
**Pull request by zvolchak**

### CODE REVIEW

File: blueprints/30-nodes/blueprint.py [hunk 1]
```diff
@@ -251,15 +252,23 @@ def delete_node_binding(nodespec):
...
-        for node_file in glob.glob(node_image_dir + '/*'):
-            os.remove(node_file)
+        node_name = os.path.basename(node_image_dir)
+        node_build_dir = BP.config['FILESYSTEM_IMAGES'] + '/' + node_name
+
+        files_to_clean = glob.glob(node_image_dir + '/*')
+        files_to_clean.extend(glob.glob(node_build_dir + '/*')) # sys-images/$NODE$/*
+
+        for to_remove in files_to_clean:
+            file_utils.remove_target(to_remove)
```
Comment (zvolchak) [inline blueprints/30-nodes/blueprint.py]: NOTE: new code here. This will clean not only tft/images/$NODE$ folder, but also sys-images/$NODE$/*

File: utils/customize_node.py [hunk 2]
```diff
@@ -188,9 +201,11 @@ def set_apt_proxy(args):
     # insufficient.  Either way, get the proxies from os.environ.
     path = args.new_fs_dir + '/etc/apt/apt.conf.d'
     file_utils.make_dir(path)
-    path += '/00FAMproxy.conf'  # !!! FIXME: no hardcoded custom file names !!!
+    path += '/00TMMS.conf'
```
Comment (zvolchak) [inline utils/customize_node.py]: Renamed to a bit more generic name to indicate that this config was generated by "tm-manifesting'.

File: blueprints/30-nodes/blueprint.py [hunk 3]
```diff
@@ -251,15 +252,23 @@ def delete_node_binding(nodespec):
...
         node_image_dir = node_coord2image_dir(node_coord)
-        for node_file in glob.glob(node_image_dir + '/*'):
-            os.remove(node_file)
+        node_name = os.path.basename(node_image_dir)
+        node_build_dir = BP.config['FILESYSTEM_IMAGES'] + '/' + node_name
+
+        files_to_clean = glob.glob(node_image_dir + '/*')
+        files_to_clean.extend(glob.glob(node_build_dir + '/*')) # sys-images/$NODE$/*
+
+        for to_remove in files_to_clean:
```
Comment (coloroco) [inline blueprints/30-nodes/blueprint.py]: Only do this if the args.debug flag is false, during debugging it helps to leave files laying around

File: utils/customize_node.py [hunk 4]
```diff
@@ -48,16 +48,27 @@ def extract_bootfiles(args, keep_kernel=False):
...
     vmlinuz = glob.glob('%s/vmlinuz*' % (boot_dir))      # Move and keep
     initrd = glob.glob('%s/initrd.img*' % (boot_dir))    # Move and ignore
-    #also extract config-* and System.map* files installed with kernel.
+
+    is_force_keep_kernel = False
+    # 'manifest' param may not exist when building golden image.
+    if hasattr(args, 'manifest'):
+        #override 'keep_kernel' if user used 'keep_kernel' in the manifest
+        is_force_keep_kernel = getattr(args.manifest, 'keep_kernel', False)
+        keep_kernel = is_force_keep_kernel
```
Comment (coloroco) [inline utils/customize_node.py]: Why do you need two variables to track the same condition?  And where is "keep_kernel" set outside of this if statement?

File: configs/setup_golden_image.py [hunk 5]
```diff
@@ -28,7 +28,6 @@
 def customize_golden(manconfig, golden_tar, build_dir):
     '''Combine /etc/tmms settings and some hardcoded values.'''
     arg_values = {
-        'is_golden': True,      # Modifies a lot of behavior in customize_node
```
Comment (coloroco) [inline configs/setup_golden_image.py]: If this modifies a lot of behavior, why did you remove it?

File: utils/customize_node.py [hunk 6]
```diff
@@ -188,9 +201,11 @@ def set_apt_proxy(args):
     # insufficient.  Either way, get the proxies from os.environ.
     path = args.new_fs_dir + '/etc/apt/apt.conf.d'
     file_utils.make_dir(path)
-    path += '/00FAMproxy.conf'  # !!! FIXME: no hardcoded custom file names !!!
+    path += '/00TMMS.conf'
     args.apt_dot_conf = path    # for post-processing
-    if not args.is_golden:      # One final post-processing step is done later
+
+    # One final post-processing step is done later
+    if not args.hostname == 'golden':
```
Comment (coloroco) [inline utils/customize_node.py]: This is a BAD BAD BAD way to do this, it's why I pulled it out of the last round.  There are many places the state needs to be known and a string literal comparison in each is cumbersome.

File: utils/customize_node.py [hunk 7]
```diff
@@ -237,7 +254,7 @@ def add_other_mirror(args):
 
 def localhost2torms(args):
     '''Go through apt.conf.d and sources.list.d and change localhost -> torms'''
-    if args.is_golden:
+    if args.hostname == 'golden':
```
Comment (coloroco) [inline utils/customize_node.py]: Put the boolean check back.

File: utils/customize_node.py [hunk 8]
```diff
@@ -285,7 +302,7 @@ def hack_LFS_autostart(args):
     :param 'node_id': [int] expanded into R:E:N.
     :return: 'None' on success. Raise 'RuntimeError' on problems.
     """
-    if args.is_golden:
+    if args.hostname == 'golden':
```
Comment (coloroco) [inline utils/customize_node.py]: and here

File: utils/customize_node.py [hunk 9]
```diff
@@ -1196,13 +1210,8 @@ def execute(args):
...
         #------------------------------------------------------------------
-        # If making a new golden image, restore the kernel.  Otherwise
-        # assemble the bootable PXE images.
-        # ZACH: Why did I do this? Need to leave a "useful" comment next time...
-        if args.is_golden:
-            #update_status(args, 'Restoring %s to golden image' % args.vmlinuz_golden)
-            #copy_target_into(
-            #    args.vmlinuz_golden, '%s/untar/boot/' % args.build_dir)
+
+        if args.hostname == 'golden':
```
Comment (coloroco) [inline utils/customize_node.py]: and here

File: utils/customize_node.py (hunk 7 above)
Comment (zvolchak) [reply utils/customize_node.py]: Too many bloody variables.

File: utils/customize_node.py (hunk 6 above)
Comment (coloroco) [reply utils/customize_node.py]: Never compare literal strings in multiple places.  It's bad style in general and violated Python DRY specifically.   You have to keep track of every location it's done, you have to know that the magic values are (what if there are more).  Put back the boolean field the way I had it.   If it was only one instance I'd consider it but there are half a dozen and it may not be finished growing.

File: utils/customize_node.py (hunk 7 above)
Comment (coloroco) [reply utils/customize_node.py]: One variable.  Many attributes.

File: utils/customize_node.py (hunk 4 above)
Comment (zvolchak) [reply utils/customize_node.py]: One is based of "args.hostname == golden" and the other one is coming from the manifest.
And now looking through it again, I don't like it.
The idea is: you should be able to control whether or not you want to keep kernel in boot or not.
For golden builds we always want to keep kernel in boot, but extract it during the customization stage.

File: utils/customize_node.py (hunk 6 above)
Comment (zvolchak) [reply utils/customize_node.py]: Alright, I can't counter that one. I'll get back to bool. For the record, I don't like my approach nor I like the extra bool in args.
